### PR TITLE
migrating pegasus mobile yaml to json

### PR DESCRIPTION
**Pull request by mgc1194**
### What:
Switching the i18n content in `pegasus/cache/i18n` from `YAML` files to `JSON`.

### Why:
Json format is more reliable than yaml: [yaml-sucks](https://github.com/cblp/yaml-sucks)
Eventually, we want to switch all i18n related yaml files into json format.

### This PR Does:
- Removes `YAML` files from `pegasus/cache/i18n`
- Add `JSON` files with existing content to `pegasus/cache/i18n`
- Modifies the `pegasus/mobile/sync_in` method to read a json file instead of yaml
- Modifies the `pegasus/mobile/sync_out` method to distribute `JSON` file instead of `YAML`
- Updates the method `parse_file` in `i18n_script_utils` to inform better about exceptions accused while parsing files.
- Add unit test coverage for new behaviour.
- Includes an i18n sync of the new `pegasus/mobile` content where translations are provided by Crowdin's translation memory of the previous `YAML` files.
- Updated pegasus/src/env and middleware/varnish_environment to load translations from `JSON` files instead of `YAML`

## Testing story
Screenshot 2024-07-31 at 12 00 49

Screenshot 2024-07-31 at 13 04 48

## Deployment strategy
- This PR should be deployed after https://github.com/code-dot-org/code-dot-org/pull/59678 is deployed

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labelled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -18,6 +18,7 @@ dashboard/app/assets/** filter=lfs diff=lfs merge=lfs -text
 dashboard/config/locales/** filter=lfs diff=lfs merge=lfs -text
 dashboard/public/** filter=lfs diff=lfs merge=lfs -text
 i18n/locales/** filter=lfs diff=lfs merge=lfs -text
+pegasus/cache/i18n/** filter=lfs diff=lfs merge=lfs -text
```
Comment (mgc1194) [inline .gitattributes:21]: Adding the i18n files to `Git LFS`

File: .gitattributes [hunk 2]
```diff
@@ -38,6 +39,7 @@ dashboard/app/assets/javascripts/** !text !filter !merge !diff
 dashboard/app/assets/stylesheets/** !text !filter !merge !diff
 dashboard/config/locales/**/*en.yml !text !filter !merge !diff
 i18n/locales/source/** !text !filter !merge !diff
+pegasus/cache/i18n/en-US.json !text !filter !merge !diff
```
Comment (mgc1194) [inline .gitattributes:42]: Excluding the English file from `Git LFS` so PRs display changes to the English strings.

File: bin/i18n/crowdin/etags/pegasus/mobile.codeorg.json
Comment (mgc1194) [inline bin/i18n/crowdin/etags/pegasus/mobile.codeorg.json:1]: Etag updates after running the sync down.

File: bin/i18n/i18n_script_utils.rb [hunk 3]
```diff
@@ -321,10 +321,18 @@ def self.sanitize_data_and_write(data, dest_path)
...
+    return JSON.parse(file_content) if json_file?(file_path)
+    return YAML.safe_load(file_content) if yaml_file?(file_path)
 
     raise "do not know how to parse file #{file_path.inspect}"
+  rescue Errno::ENOENT => exception
+    raise "File not found: #{file_path.inspect} - #{exception.message}"
+  rescue JSON::ParserError => exception
+    raise "JSON parsing error in file #{file_path.inspect} - #{exception.message}"
+  rescue Psych::SyntaxError => exception
+    raise "YAML parsing error in file #{file_path.inspect} - #{exception.message}"
```
Comment (mgc1194) [inline bin/i18n/i18n_script_utils.rb:335]: These changes are completely arbitrary, I just wanted to produce more detailed Error logs, and handle `File.read` errors separately from parsing errors.

File: bin/i18n/resources/pegasus/mobile/sync_in.rb [hunk 4]
```diff
@@ -10,9 +10,17 @@ module Pegasus
       module Mobile
         class SyncIn < I18n::Utils::SyncInBase
           def process
-            # TODO: fix `i18n/locales/source/pegasus/mobile.yml` instead of the original file `pegasus/cache/i18n/en-US.yml`
-            I18nScriptUtils.fix_yml_file(ORIGINAL_I18N_FILE_PATH)
```
Comment (mgc1194) [inline bin/i18n/resources/pegasus/mobile/sync_in.rb:14]: We are not using YALML anymore therefore we don't need to fix the file.

**Review by artem-vavilov — APPROVED** [review]
Good job! 🙌🏻

File: bin/i18n/resources/pegasus/mobile/sync_in.rb [hunk 5]
```diff
@@ -10,9 +10,17 @@ module Pegasus
...
+            if i18n_data && i18n_data['en-US']
+              I18nScriptUtils.sanitize_data_and_write(i18n_data['en-US'], I18N_SOURCE_FILE_PATH)
+            else
+              # Handle the case where i18n_data is nil or does not contain 'en-US'
+              handle_missing_or_invalid_data
+            end
+          end
+
+          def handle_missing_or_invalid_data
+            raise "Error: i18n data is missing or invalid"
```
Comment (artem-vavilov) [inline bin/i18n/resources/pegasus/mobile/sync_in.rb:23]: Can we provide more information to identify which locale is missing its i18n data?

File: bin/i18n/resources/pegasus/mobile/sync_in.rb [hunk 6]
```diff
@@ -10,9 +10,17 @@ module Pegasus
...
-            # TODO: fix `i18n/locales/source/pegasus/mobile.yml` instead of the original file `pegasus/cache/i18n/en-US.yml`
-            I18nScriptUtils.fix_yml_file(ORIGINAL_I18N_FILE_PATH)
-            I18nScriptUtils.copy_file(ORIGINAL_I18N_FILE_PATH, I18N_SOURCE_FILE_PATH)
+            i18n_data = I18nScriptUtils.parse_file(ORIGINAL_I18N_FILE_PATH)
+            if i18n_data && i18n_data['en-US']
+              I18nScriptUtils.sanitize_data_and_write(i18n_data['en-US'], I18N_SOURCE_FILE_PATH)
+            else
+              # Handle the case where i18n_data is nil or does not contain 'en-US'
+              handle_missing_or_invalid_data
+            end
```
Comment (artem-vavilov) [inline bin/i18n/resources/pegasus/mobile/sync_in.rb:19]: Just an example, no need to change anything:
```suggestion
            i18n_data = I18nScriptUtils.parse_file(ORIGINAL_I18N_FILE_PATH).try(:[], 'en-US')
            return handle_missing_or_invalid_data unless i18n_data

            I18nScriptUtils.sanitize_data_and_write(i18n_data, I18N_SOURCE_FILE_PATH)
```

File: bin/i18n/resources/pegasus/mobile/sync_in.rb [hunk 7]
```diff
@@ -10,9 +10,17 @@ module Pegasus
...
           def process
-            # TODO: fix `i18n/locales/source/pegasus/mobile.yml` instead of the original file `pegasus/cache/i18n/en-US.yml`
-            I18nScriptUtils.fix_yml_file(ORIGINAL_I18N_FILE_PATH)
-            I18nScriptUtils.copy_file(ORIGINAL_I18N_FILE_PATH, I18N_SOURCE_FILE_PATH)
+            i18n_data = I18nScriptUtils.parse_file(ORIGINAL_I18N_FILE_PATH)
+            if i18n_data && i18n_data['en-US']
+              I18nScriptUtils.sanitize_data_and_write(i18n_data['en-US'], I18N_SOURCE_FILE_PATH)
+            else
+              # Handle the case where i18n_data is nil or does not contain 'en-US'
+              handle_missing_or_invalid_data
```
Comment (artem-vavilov) [inline bin/i18n/resources/pegasus/mobile/sync_in.rb:18]: I think the method name is clear enough that there's no need to explain its meaning in a comment.

**Review by bethanyaconnor — APPROVED** [review]
Can we still do the namespace pattern in json? ie

```
"my_launch" {
  "str1": "String 1",
  ....
}
```
?

Also fyi @kelbyhawn about this change. We haven't been using `pegasus/cache/i18n/en-US.json` for translations because of the Dropbox sync so I'm not sure how much we'll really change our process at this point, but it would be good to understand any limitations!

**Comment by kelbyhawn** [comment]
Ooo exciting! Could we start using this file instead of [hourofcode.com/i18n/en.yml](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/sites.v3/hourofcode.com/i18n/en.yml) for code.org site specific strings going forward? 

And does this still use the `i18n.t()` method?

**Comment by mgc1194** [comment]
> Can we still do the namespace pattern in json? ie
> 
> ```
> "my_launch" {
>   "str1": "String 1",
>   ....
> }
> ```
> 
> ?
> 
> Also fyi @kelbyhawn about this change. We haven't been using `pegasus/cache/i18n/en-US.json` for translations because of the Dropbox sync so I'm not sure how much we'll really change our process at this point, but it would be good to understand any limitations!

Yes absolutely, hierarchy for namespaces can be used.

**Comment by mgc1194** [comment]
> Ooo exciting! Could we start using this file instead of [hourofcode.com/i18n/en.yml](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/sites.v3/hourofcode.com/i18n/en.yml) for code.org site specific strings going forward?
> 
> And does this still use the `i18n.t()` method?

Yes, now strings can be added to this file could be used the same way we've been using `hourofcode.com/i18n/en.yml` and the method will be `I18n.t()`.

**Pull request merged**